libtransmission/blocklist.cc

```diff
@@ -445,9 +445,32 @@ size_t BlocklistFile::setContent(char const* filename)
         return {};
     }
 
-    size_t keep = 0; // index in ranges
+    //seperate before sorting
+    auto ipv4_ranges = std::vector<AddressRange>{};
+    auto ipv6_ranges = std::vector<AddressRange>{};
+
+    for (auto const& range : ranges)
+    {
+        if (range.begin_ == 0 && range.end_ == 0)
+        {
+            // IPv6
+            ipv6_ranges.emplace_back(range);
+        }
+        else
+        {
+            ipv4_ranges.emplace_back(range);
+        }
+    }
+
+    std::sort(std::begin(ipv4_ranges), std::end(ipv4_ranges), BlocklistFile::compareAddressRangesByFirstAddress);
+    std::sort(std::begin(ipv6_ranges), std::end(ipv6_ranges), BlocklistFile::compareAddressRangesByFirstAddress);
 
-    std::sort(std::begin(ranges), std::end(ranges), BlocklistFile::compareAddressRangesByFirstAddress);
+    // combine sorted
+    ranges.clear();
+    ranges.insert(ranges.end(), ipv4_ranges.begin(), ipv4_ranges.end());
+    ranges.insert(ranges.end(), ipv6_ranges.begin(), ipv6_ranges.end());
```

Comment:
Subsequent lookups in `hasAddress()` happen depending on address type. I don't think the idea of holding both IPv4 and IPv6 addresses in the same vector was good to begin with, it just adds more complexity, higher memory use, and reduced performance.

I'd split `rules_` in two variables, i.e. `rules_ipv4_` and `rules_ipv6_`, which you have already started preparing here. This in turn means that each variable could use
* its own type-specific `AddressRange` structure to reduce memory use
* its own comparator: we already have `AddressRange::compareIPv4AddressToRange()` and `AddressRange::compareIPv6AddressToRange()`, each could go to its own structure (as `operator<()`), just need to make them proper C++ comparators by switching from `void*` to proper type, and using `std::binary_search()` (C++) instead of `std::bsearch()` (C); since C++ comparators require you to return `bool` and not `int`, this will also mean one less `memcmp()` in IPv6 comparison (although I'm not sure why there're two right now); no need in additional `compareAddressRangesByFirstAddress()`
* its own file to write to and load from, if it makes sense

If there's a way to make this generic (e.g. using templates), I'm all for it.

Reply:
I agree with all this, it makes total sense to seperate ipv4 and ipv6 addresses.
but perhaps it would be better suited to a seperate PR after this fix is merged

Reply:
In its current form, this fix helps little because all these assertions on separated parts mean nothing when it comes to `hasAddress()` which searches in a merged vector. Even if unrelated structure parts are zeroed out, it still doesn't make the vector suitable for binary search; the final merge result contains IPv4 address ranges followed by IPv6 address ranges, which would mean that
* IPv6 part of IPv4 addresses is zeroed out; when IPV4 address ranges are treated as IPv6 address ranges (during IPv6 address lookup) they are less than any other [actual] IPv6 address range, making the vector sorted in the eyes of `std::bsearch()` (because IPv4 address ranges come first, and should be less than anything else following them)
* IPv4 part of IPv6 addresses is zeroed out; when IPv6 address ranges are treated as IPv4 address ranges (during IPv4 address lookup) they are less than any other [actual] IPv4 address range, making the vector unsorted in the eyes of `std::bsearch()` (because IPv6 address ranges come last, and should be greater than anything else preceding them)

Supplying unsorted sequences to `std::bsearch()` (and `std::binary_search()` for that matter) will yield unpredictable results. It may take time to find the search key or the blocklist for this to happen, but it will happen sooner or later.

Not to mention the merging issue at the position between the last IPv4 address range and the first IPv6 address range in the joined sequence: upon encountering its first IPv6 address range, the loop treats the last IPv4 address range as IPv6 address range, which may lead to them being (wrongfully) merged together.

Reply:
I’m looking into this, but work is calling right now. So time is limited

Reply:
I’ve had a play around with this, but I can’t get my head around the syntax for templates. C++ is not a language I am familiar with in any depth.

perhaps someone else wants to finish the migration in a new PR